Clerk Elements (Beta) - Initial docs (#890)
clerk/clerk-docs: docs/elements/overview.mdx

---
description: Learn how to use Clerk Elements to build custom UIs on top of Clerk's APIs without having to manage the underlying logic.
---

# Clerk Elements (Beta)

<Callout type="warning">
Clerk Elements is currently in beta and **not yet recommended for production use**.

If you have any feedback, please reach out to [beta-elements@clerk.dev](mailto:beta-elements@clerk.dev) or head over to the [GitHub Discussion](#TODO).
</Callout>

Clerk Elements is a library of unstyled, composable components that can be used to build custom UIs on top of Clerk's APIs without having to manage the underlying logic.

## Why use Clerk Elements?

You should use Clerk Elements if you want a deeper level of control and customization of the styles and layout of your UI while using Clerk's APIs. For example, if [the appearance prop](/docs/components/customization/overview) does not meet your needs, Clerk Elements might be for you. That said, you can also use Clerk Elements alongside the prebuilt components.

Two very important goals of Clerk Elements are:

> - **Component-first** - Make it as easy to build custom UIs with Clerk as it is with Clerk's drop-in prebuilt components. Clerk Elements handles the underlying business logic for you and provides a curated library of components without sacrificing on best practices or features.

Author: Yep, this is accurate! Currently Clerk Elements is used entirely with components, and no state / event listeners / effects need to be created or managed by the user. So it's entirely just markup and styling.

> - **A little bit of magic** - Use the web platform and build best-in-class components for building great authentication flows. Baked-in to the components are little bits of magic, like the fully accessible segmented one-time-passcode (OTP) input, or instant password validation during sign up.

Author: I generally agree with you! However, in the absence of a dedicated marketing page for Elements I think it's important to add some color commentary on this page to entice people and communicate why we think it's a valuable piece of our product offering.

This is why I had it in a separate "Highlights" section previously.

- **With Clerk's prebuilt components** - Continue using Clerk's prebuilt components while customizing the flows you care most about.

## Getting started

Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added. If your Next.js application is already using Clerk, make sure to [upgrade to Core 2](/docs/upgrade-guides/core-2/nextjs). If you're starting from scratch, follow the [Next.js quickstart](/docs/quickstarts/nextjs) before proceeding.

To get started, install the Clerk Elements package:

<CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
  ```sh filename="terminal"
  npm install @clerk/elements
  ```

  ```sh filename="terminal"
  yarn add @clerk/elements
  ```

  ```sh filename="terminal"
  pnpm add @clerk/elements
  ```
</CodeBlockTabs>

Once you have your project set up, you can start building custom UIs with Clerk Elements using Clerk's guides and examples. For example, to use Clerk Elements to build a custom sign-in flow, you can explore:

> - [Build a sign-in flow](/docs/elements/guides/sign-in)

Reviewer: The text above says `For example, to use Clerk Elements to build a custom _sign-in flow_, you can explore`

so we wouldn't mention sign-up flow here

---

> description: Clerk Elements is a library of unstyled, composable components that can be used to build completely custom UIs on top of Clerk's APIs, without having to manage the underlying logic.

Reviewer 2: ```suggestion
description: Learn how to use Clerk Elements to build custom UIs on top of Clerk's APIs without having to manage the underlying logic.
```

---

# Clerk Elements (Beta)

<Callout type="warning">

> Clerk Elements is currently in beta. While it is **not** yet recommended for production use, it would be much appreciated if you gave it a try.

Reviewer 2: ```suggestion
Clerk Elements is currently in beta and **not yet recommended for production use**.
```

> Clerk Elements is currently in beta. It's **not** recommended to use it in production just yet, but it would be much appreciated if you give it a try.

Reviewer 3: ```suggestion
Clerk Elements is currently in beta. We are not ready to recommended using it in production just yet, but it would be much appreciated if you gave it a try.
```

If you have any feedback, please reach out to [beta-elements@clerk.dev](mailto:beta-elements@clerk.dev) or head over to the [GitHub Discussion](#TODO).
</Callout>

> Clerk Elements is a library of unstyled, composable components that can be used to build completely custom UIs on top of Clerk's APIs, without having to manage the underlying logic.

Reviewer 2: ```suggestion
Clerk Elements is a library of unstyled, composable components that can be used to build custom UIs on top of Clerk's APIs without having to manage the underlying logic.
```

## Getting started

> Clerk Elements currently only works with Next.js and [Clerk Core 2](https://clerk.com/changelog/2024-04-19). As it gets closer to a stable release, support for additional frameworks will be added. If your Next.js application is already using Clerk, make sure to [upgrade to Core 2](/docs/upgrade-guides/core-2/nextjs). If you're starting from scratch, follow the [Next.js quickstart](/docs/quickstarts/nextjs) before proceeding.

Reviewer 2: Does it matter which version of Next.js?

Reviewer 4: Yeah, one needs to use `^13.5.4 || ^14.0.3`

Author: This is already specified in our Next.js SDK, so I don't think we need to explicitly call it out here.

Reviewer 2: Ah, okay, that makes sense

To get started, install the Clerk Elements package:

<CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
  ```sh filename="terminal"
  npm install @clerk/elements
  ```

  ```sh filename="terminal"
  yarn add @clerk/elements
  ```

  ```sh filename="terminal"
  pnpm add @clerk/elements
  ```
</CodeBlockTabs>

> Once you've got your project set up, you can start building custom UIs with Clerk Elements. The following guides will show you how to build your own custom flows:

Reviewer 3: ```suggestion
Once you have your project set up, you can start building custom UIs with Clerk Elements. The following guides will show you how to build your own custom flows:
```

> Once you have your project set up, you can start building custom UIs with Clerk Elements. The following guides will show you how to build your own custom flows:
> - [Build a sign-in flow](/docs/elements/guides/sign-in)
> - [Build a sign-up flow](/docs/elements/guides/sign-up)
> 
> If you want to see what's possible with Clerk Elements, check out these pre-built examples from the Clerk team:
> - [Sign-in examples](/docs/elements/examples/sign-in)
> - [Sign-up examples](/docs/elements/examples/sign-up)
> 
> Finally, to learn about the available components and how to use them, check out the component reference pages:
> - [Common components](/docs/elements/reference/common)
> - [Sign-in components](/docs/elements/reference/sign-in)
> - [Sign-up components](/docs/elements/reference/sign-up)

Reviewer 2: ```suggestion
Once you have your project set up, you can start building custom UIs with Clerk Elements using Clerk's guides and examples. For example, to use Clerk Elements to build a custom sign-in flow, you can explore:

- [Build a sign-in flow](/docs/elements/guides/sign-in)
- [Sign-in examples](/docs/elements/examples/sign-in)
- [Sign-in components](/docs/elements/reference/sign-in)
```

> _With the beta release, only sign-up and sign-in flows are supported. Support for building the rest of Clerk's prebuilt components with Elements is actively being worked on._

Reviewer 2: ```suggestion
<Callout type="info">
With the beta release, only sign-up and sign-in flows are supported. Support for building the rest of Clerk's prebuilt components with Elements is actively being worked on.
</Callout>
```
